lib/utils/normalization.ts

```diff
@@ -15,17 +15,31 @@ limitations under the License.
 */
 
 import type { BalenaSDK } from 'balena-sdk';
-import _ = require('lodash');
 import { ExpectedError } from '../errors';
 
-export function normalizeUuidProp(
-	params: { [key: string]: any },
-	propName = 'uuid',
+const onlyNumericalRegex = /^[0-9]+$/;
+
+export async function disambiguateAppParam(
+	balena: BalenaSDK,
+	appIdentifier: string,
 ) {
-	if (typeof params[propName] === 'number') {
-		params[propName] =
-			params[propName + '_raw'] || _.toString(params[propName]);
+	// Only need to disambiguate if the appIdentifier is all-numerical
+	if (onlyNumericalRegex.test(appIdentifier)) {
+		// Check if an app exists with a numerical-name
+		try {
+			await balena.models.application.get(appIdentifier, {
+				$select: 'id',
+			});
+		} catch (e) {
+			if (e.name !== 'BalenaApplicationNotFound') {
+				throw e;
+			}
+			// No numerical-name app exists, so return id cast to number.
+			return Number(appIdentifier);
+		}
 	}
+
+	return appIdentifier;
```

Comment:
I suggest not discarding a database query that actually found the application -- as another similar call to `balena.models.application.get()` would follow this return statement, for the same `appIdentifier`.

Instead of having a function like `disambiguateAppParam` whose purpose is just to decide whether the app identifier should be a number or a string, we could have a CLI's own `getApplication` method (perhaps in `patterns.ts`) whose purpose was to get and return the application (with any given options) by first attempting to fetch a `looksLikeInteger()` identifier as a name string, and if this failed, then tried again as a database ID number. Along the lines of:

```javascript
// lib/utils/patterns.ts

export async function getApplication(
	sdk: BalenaSDK,
	nameOrSlugOrId: string | number,
	options?: PineOptions<Application>,
): Promise<Application> {
	if (looksLikeInteger(nameOrSlugOrId)) {
		try {
			return sdk.models.application.get(nameOrSlugOrId, options);
		} catch (err) {
			if (instanceof(err, BalenaApplicationNotFound)) {
				nameOrSlugOrId = parseInt(nameOrSlugOrId, 10);
			} else {
				throw err;
			}
		}
	}
	return sdk.models.application.get(nameOrSlugOrId, options);
}
```
Then in `lib/commands/app/index.ts`, we'd call this `getApplication` function instead of `balena.models.application.get()`.

Reply:
Yeah this is a better approach.   

I've pushed a variation that, if looksLikeInteger, it tests for the integer ID first, then the name.   The benefit of this is that in situations where we have e.g.:  
NAME   ID 
1234     11111
test        1234
We can still access them both via ID, (otherwise the second would not be accessible by ID), and we can still access the second by name, using the slug `sl/1234`.

Let me know what you think about this inversion, I think it's slightly better, but don't have strong feelings.